File: 5/5-2.pl
#!/usr/bin/perl -wl
use strict;
use warnings FATAL => 'all';
use Data::Dumper;

my $fname = shift || '-';
open my $fh, '<', $fname or die;
my (%rules, $updates);
{
    local $/ = "\n\n";
    %rules = map {$_, 1} split /\n/, scalar <$fh>;
    $updates = scalar <$fh>;
}

my $total;
for my $update (split /\n/, $updates) {
    my @pages = split /,/, $update;
    my $ok = 1;
    for my $i (0..$#pages-1) {
        for my $j ($i+1..$#pages) {
            my $rule = join '|', $pages[$i],$pages[$j];
            $ok = $rules{$rule};
            last unless $ok;
        }
        last unless $ok;
    }
    if (!$ok) {
        my @order;
        my %pendingset = map {$_,1} @pages;
        my $numpages = scalar @pages;
        for (0..$numpages-1) {
            my @pending = keys %pendingset;
            for my $candidateIndex (0..$#pending) {
                my $candidate = $pending[$candidateIndex];
                my @others = grep {$_ != $candidate} @pending;
                if (!grep {!exists $rules{join '|', $candidate, $_}} @others) {
                    push @order, $candidate;
                    delete $pendingset{$candidate};
                    last;
                }
            }
        }
        die "No answer for $update" if %pendingset;
        $total += $order[($#order+1)/2];
    }
}

print $total;

File: 5/5-2.py
import sys

if len(sys.argv) < 2:
    file = sys.stdin
else:
    file = open(sys.argv[1], 'rt')

rules = set()
for line in file:
    line = line.rstrip()
    if not line:
        break
    rules.add(line)

updates = list()
for line in file:
    updates.append(line.rstrip())

total = 0

for update in updates:
    pages = update.split(',')
    ok = True
    for i in range(0, len(pages) - 1):
        for j in range(i + 1, len(pages)):
            rule = '|'.join([pages[i], pages[j]])
            ok = rule in rules
            if not ok:
                break
        if not ok:
            break

    if not ok:
        order = list()
        pendingset = set(pages)
        for _ in range(0,len(pages)):
            pending = list(pendingset)
            for candidateIndex in range(0,len(pending)):
                candidate = pending[candidateIndex]
                others = list(pending)
                others.remove(candidate)
                ok = True
                for other in others:
                    rule = '|'.join([candidate, other])
                    ok = rule in rules
                    if not ok:
                        break
                if ok:
                    order.append(candidate)
                    pendingset.remove(candidate)
                    break
        if pendingset:
            print(f"No answer for {update}")
        else:
            total += int(order[((len(order) + 1) // 2)-1])

print(total)
